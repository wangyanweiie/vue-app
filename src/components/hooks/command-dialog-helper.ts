import {
    type AppContext,
    type Component,
    type ComponentPublicInstance,
    createVNode,
    getCurrentInstance,
    render,
    type VNode,
} from 'vue';

interface Options {
    visible?: boolean;
    onClose?: () => void;
    appendTo?: HTMLElement | string;
    [key: string]: unknown;
}

interface CommandComponent {
    (options: Options): VNode;
    close: () => void;
}

/**
 * @description 获取挂载位置
 * @param props 自定义选项
 * @returns 挂载位置
 */
const getAppendToElement = (props: Options): HTMLElement => {
    console.log('props', props);

    let appendTo: HTMLElement | null = document.body;

    if (props.appendTo) {
        // 若 props.appendTo 为选择器时，通过 querySelector 获取到该元素
        if (typeof props.appendTo === 'string') {
            appendTo = document.querySelector<HTMLElement>(props.appendTo);
        }

        // 若 props.appendTo 为 html 元素，直接赋值
        if (props.appendTo instanceof HTMLElement) {
            appendTo = props.appendTo;
        }

        // 否则赋值为 body
        if (!(appendTo instanceof HTMLElement)) {
            appendTo = document.body;
        }
    }

    return appendTo;
};

/**
 * @description 渲染虚拟节点
 * @param Component 要传入的组件
 * @param props 自定义选项
 * @param container 挂载位置
 * @param appContext 上下文
 * @returns 虚拟节点
 */
const handleInitInstance = <T extends Component>(
    Component: T,
    props: Options,
    container: HTMLElement,
    appContext: AppContext | null = null,
) => {
    // 渲染函数 <==> h()
    const vNode = createVNode(Component, props);
    vNode.appContext = appContext;

    // 编程式地创建组件虚拟 DOM 树的函数
    render(vNode, container);

    // 挂载组件
    getAppendToElement(props).appendChild(container);
    return vNode;
};

/**
 * @description 命令式组件
 * @param Component 要传入的组件
 * @returns 命令式组件实例
 */
const useCommandComponent = <T extends Component>(Component: T): CommandComponent => {
    // 获取当前组件的实例、上下文
    const appContext = getCurrentInstance()?.appContext;

    /**
     * 由于命令式组件无法获取当前组件树的 injection，
     * 可以将当前组件树的 provides 与 appContext.provides 合并，
     * 这样传入的弹窗组件就可以顺利的获取到 app 和当前组件树的 provides
     * Reflect.set() <==> lodash.set()
     */
    if (appContext) {
        const currentProvides = (getCurrentInstance() as any)?.provides;
        Reflect.set(appContext, 'provides', { ...appContext.provides, ...currentProvides });
    }

    // 创建容器
    const container = document.createElement('div');

    // 关闭方法，删除容器节点
    const close = () => {
        render(null, container);
        container.parentNode?.removeChild(container);
    };

    // 声明组件实例
    const commandComponent = (options: Options): VNode => {
        if (!Reflect.has(options, 'visible')) {
            options.visible = true;
        }

        if (typeof options.onClose !== 'function') {
            options.onClose = close;
        } else {
            const originOnClose = options.onClose;
            options.onClose = () => {
                originOnClose();
                close();
            };
        }

        const vNode = handleInitInstance<T>(Component, options, container, appContext);
        const vm = vNode.component?.proxy as ComponentPublicInstance<Options>;

        // 将选项中的属性赋值到创建的节点上
        for (const prop in options) {
            // Reflect.has() <==> lodash.set()
            if (!Reflect.has(vm.$props, prop)) {
                // ComponentPublicInstance 组件类型，只会包含所有组件都共享的属性
                vm[prop as keyof ComponentPublicInstance] = options[prop];
            }
        }

        return vNode;
    };

    commandComponent.close = close;
    return commandComponent;
};

export { type CommandComponent, useCommandComponent };
