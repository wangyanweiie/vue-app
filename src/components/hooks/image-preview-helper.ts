import { createVNode, render } from 'vue';
import { ElImageViewer, type ImageViewerProps } from 'element-plus';

type PreviewOption = Partial<ImageViewerProps>;

/**
 * @description 图片预览
 * @param option { PreviewOption } 图片预览的选项
 */
export function previewImage(option: PreviewOption) {
    // 创建了一个 div 元素作为容器，用于渲染预览组件
    const container = document.createElement('div');

    // 使用 createVNode 创建了一个 ElImageViewer 组件实例 vm
    const vm = createVNode(ElImageViewer, {
        ...option,
        onClose() {
            // 监听 close 事件删除节点
            render(null, container);
        },
    });

    // 使用 render 方法将 vm 渲染为真实的节点，并将其挂载到之前创建的容器中
    render(vm, container);

    // 添加到 body 页面上
    document.body.appendChild(container.firstElementChild!);
}
