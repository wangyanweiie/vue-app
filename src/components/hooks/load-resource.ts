/**
 * 异步加载 JavaScript 文件。
 * @param src 要加载的 JavaScript 文件的 URL。
 * @param clearCache 是否在 src 中添加时间戳参数以清除缓存，默认为 false。
 */
export function loadJS(src: string, clearCache = false) {
    // 创建一个新的 Promise，通过动态创建 script 元素并设置其属性来加载 JavaScript 文件
    new Promise<HTMLScriptElement>((resolve, reject) => {
        // 创建 script 元素
        const script = document.createElement('script', {});
        script.type = 'text/javascript';

        // 处理清除缓存逻辑
        if (clearCache) {
            script.src = src + '?t=' + new Date().getTime();
        } else {
            script.src = src;
        }

        // 统一的事件处理函数，以减少代码重复
        const handleEvent = (event: Event) => {
            // 根据事件类型执行相应操作
            if (event.type === 'load') {
                resolve(script);
            } else if (event.type === 'error') {
                const errorMsg = `Failed to load script: ${src}`;
                reject(new Error(errorMsg));
            }
        };

        // 监听脚本加载成功或失败事件
        if (script.addEventListener) {
            script.addEventListener('load', handleEvent);
            script.addEventListener('error', handleEvent);
        }

        // 将 script 元素添加到文档头部
        document.head.appendChild(script);
    });
}

/**
 * 动态加载 CSS 文件到当前页面
 * @param src CSS 文件的 URL 地址
 */
export function loadCSS(src: string) {
    // 创建一个 link 元素
    const file = document.createElement('link');

    // 设置 link 元素的属性为 CSS 样式表
    file.setAttribute('rel', 'stylesheet');
    file.setAttribute('type', 'text/css');

    // 设置 CSS 文件的路径
    file.setAttribute('href', src);

    // 添加错误处理
    file.onerror = () => {
        console.error('Failed to load CSS file:', src);
    };

    // 将 link 元素添加到文档头部
    document.head.appendChild(file);
}
