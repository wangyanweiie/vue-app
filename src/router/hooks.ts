import type { RouteRecordRaw } from 'vue-router';

/**
 * 递归处理 component 字符串
 * @param json 路由字符串
 * @returns 路由列表
 */
export function processComponentStrings(routes: any[]): RouteRecordRaw[] {
    routes.forEach(route => {
        // 处理当前路由的 component
        if (typeof route.component === 'string' && route.component.startsWith('() =>')) {
            // 移除字符串的引号,并将字符串转换为实际的箭头函数
            route.component = new Function(`return ${route.component}`)();
        }

        // 递归处理子路由
        if (route.children && route.children.length) {
            processComponentStrings(route.children);
        }
    });

    return routes;
}
