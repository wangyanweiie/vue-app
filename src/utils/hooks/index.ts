import type { Options } from './interface';

/**
 * 校验是否为数字类型（正数/负数/整数/小数）
 * @param value 字符串
 */
export function checkNumberFormat(value: string) {
    if (!value) {
        return;
    }

    // 正则表达式
    const reg = /^[+-]?(\d|([1-9]\d+))(\.\d+)?$/;
    const flag = reg.test(value);

    if (flag) {
        return value;
    } else {
        return null;
    }
}

/**
 * 强制保留小数位方法
 * @param value 要处理的数据
 * @param precision 小数位数
 */
export function keepDecimalPrecision(value: number | string, precision: number) {
    if (!value) {
        return;
    }

    // 将数字转换为字符串
    let res = String(value);

    // 小数点的索引值
    let posDecimalIndex = res.indexOf('.');

    // 当整数，即 posDecimalIndex = -1 时，拼接小数点
    if (posDecimalIndex === -1) {
        posDecimalIndex = res.length;
        res += '.';
    }

    // 当数字的长度 < 小数点索引 + precision 时，用 0 补全
    while (res.length <= posDecimalIndex + precision) {
        res += '0';
    }

    return res;
}

/**
 * 将枚举转换为 options
 * @param enumeration 枚举
 */
export function transformEnumToOptions(enumeration: Record<string, string | number>): Options[] {
    // Object.entries 返回给定对象自身可枚举属性的键值对数组
    const list = Object.entries(enumeration);
    const transList = list
        .map(([label, value]) => {
            return {
                label,
                value: value as number,
            };
        })
        .slice(list.length / 2);

    return transList;
}

/**
 * 将 '-' 拼接字符串改为驼峰格式
 * @param str 要转换的字符串
 * @param type 要转换的驼峰格式
 */
export function handleToHumpFormat(str: string, type: 'min' | 'max') {
    switch (type) {
        // 小驼峰格式：x-table ==> xTable
        case 'min': {
            const reg = /[-_](\w)/g;
            return str.replace(reg, (initial, item) => {
                return item.toUpperCase();
            });
        }

        // 大驼峰格式：x-table ==> XTable
        case 'max': {
            const arr = str.split('-');
            const res = arr.map(item => `${item[0].toUpperCase()}${item.slice(1, item.length)}`).join('');
            return res;
        }
    }
}

/**
 * 生成 uuid
 */
export function guid() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        const r = (Math.random() * 16) | 0,
            v = c == 'x' ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
}
