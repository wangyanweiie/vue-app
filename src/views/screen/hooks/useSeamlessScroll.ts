import { onMounted, shallowRef, type Ref } from 'vue';
import anime from 'animejs/lib/anime.es.js';

export type OptionsType = {
    direction?: 'horizontal' | 'vertical';
    gap?: number;
    duration?: number;
};

/**
 * @description 无缝滚动
 * @param listRef 滚动列表的 ref
 * @param options.direction 滚动方向
 * @param options.gap 滚动间隔
 * @param options.duration 动画持续时间
 */
export function useSeamlessScroll(listRef: Ref<HTMLElement | null>, options: OptionsType = {}) {
    const { direction = 'horizontal', gap = 10, duration = 20000 } = options;
    const animation = shallowRef<ReturnType<typeof anime> | null>(null);

    /**
     * @description 初始化
     */
    function init() {
        const isHorizontal = direction === 'horizontal';
        const translateKey = isHorizontal ? 'translateX' : 'translateY';
        const transKey = isHorizontal ? 'x' : 'y';

        const children = listRef.value?.children || [];
        if (!children.length) {
            return;
        }

        const firstEl = children[0] as HTMLElement;
        const firstDiff = (isHorizontal ? firstEl.offsetWidth : firstEl.offsetHeight) + gap;

        // 默认将 list 元素向左或向上移动一个 item 的距离
        listRef.value!.style.transform = `${translateKey}(-${firstDiff}px)`;

        let total = 0; // 总宽或总高
        const transList: Record<string, number>[] = [];

        // 设置初始位置
        anime.set(children, {
            [translateKey]: (el: HTMLElement, i: number) => {
                const distance = (isHorizontal ? el.offsetWidth : el.offsetHeight) + gap;
                total += distance;

                transList[i] = { [transKey]: total - distance };
            },
        });

        // 设置 list 容器的宽或高
        listRef.value!.style[isHorizontal ? 'width' : 'height'] = total + 'px';

        // 添加动画
        animation.value = anime({
            targets: transList,
            duration,
            easing: 'linear',
            direction: isHorizontal ? undefined : 'reverse',
            [transKey]: `+=${total}`,
            loop: true,
            update: () => {
                anime.set(children, {
                    [translateKey]: (el: HTMLElement, i: number) => {
                        return transList[i][transKey] % total;
                    },
                });
            },
        });
    }

    /**
     * @description 暂停动画
     */
    function pause() {
        animation.value!.pause();
    }

    /**
     * @description 播放
     */
    function play() {
        animation.value!.play();
    }

    onMounted(() => {
        init();
    });

    return {
        listRef,
        animation,
        pause,
        play,
    };
}
