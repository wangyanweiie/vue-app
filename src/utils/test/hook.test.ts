import { describe, it } from 'vitest';

import { checkNumberFormat, handleToHumpFormat, keepDecimalPrecision, transformEnumToOptions } from '../hooks';

enum STATUS {
    '在用' = 1,
    '停用' = 2,
}

const statusList = [
    {
        label: '在用',
        value: 1,
    },
    {
        label: '停用',
        value: 2,
    },
];

/**
 * concurrent 并发运行
 * 还可以将 .skip、.only 以及 .todo 用于并发测试套件和测试用例
 */
describe('hooks', () => {
    it.concurrent('是否为数字格式', async ({ expect }) => {
        // expect(checkNumberFormat('3.1415')).toBe(3.1415);
        // expect(checkNumberFormat('3.14.15')).toBe(null);

        /**
         * 测试快照
         * 此测试在第一次运行时，Vitest 会创建一个快照文件，快照文件应该与代码更改一起提交，并作为代码审查过程的一部分进行审查；
         * 在随后的测试运行中，Vitest 会将执行的输出与之前的快照进行比较；若匹配，测试通过；
         * 若不匹配，要么测试运行时在你的代码中发现了应该修复的错误，要么实现已经更改，需要更新快照；
         */
        expect(checkNumberFormat('3.1415')).toMatchSnapshot();
    });

    it.concurrent('强制保留精度', async ({ expect }) => {
        expect(keepDecimalPrecision(1, 5)).toBe('1.00000');
        expect(keepDecimalPrecision(3.1415, 5)).toBe('3.14150');
    });

    it.concurrent('枚举转下拉列表', async ({ expect }) => {
        expect(JSON.stringify(transformEnumToOptions(STATUS))).toBe(JSON.stringify(statusList));
    });

    it.concurrent('字符串"-"拼接转驼峰', async ({ expect }) => {
        expect(handleToHumpFormat('x-table', 'min')).toBe('xTable');
        expect(handleToHumpFormat('x-table', 'max')).toBe('XTable');
    });
});
