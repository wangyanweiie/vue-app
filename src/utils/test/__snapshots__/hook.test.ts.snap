// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hooks > 是否为数字格式 1`] = `3.1415`;
